source/common/config/delta_subscription_state.cc

```diff
@@ -53,6 +58,31 @@ void DeltaSubscriptionState::updateSubscriptionInterest(
     names_added_.erase(r);
     names_removed_.insert(r);
   }
+  switch (mode_) {
+  case WildcardMode::Implicit:
+    if (names_removed_.find("*") != names_removed_.end()) {
```

Comment:
I just realised this implementation deviates from how wildcards are currently implemented: a subscription without any resource names is a wildcard one. A wildcard subscription stops being one when a resource name added to it. If a subscription has the last resource name removed from it, it becomes a wildcard subscription (see https://github.com/envoyproxy/envoy/blob/main/source/common/config/watch_map.cc#L23 and https://github.com/envoyproxy/envoy/blob/main/source/common/config/watch_map.cc#L50).

Another issue is that we now have wildcard-state related logic in two places; I need to think about this...

Reply:
Another difference is that currently a wildcard subscription always stays a wildcard one: added or removed resources do not affect subscriptions' behaviour after `markStreamFresh()` has been called, and there's no difference with a regular subscription at other times. 

After thinking about this, I think the way the current wildcard mode handles stream refresh might have a problem: it skips over all resource names that have been added, but doesn't remove corresponding resource states. Depending on how the server handles stream refresh (for example the server might reset its state and only send resources it considers necessary, ignoring client subscriptions), we might end up with different susbcription state on the client and the server. 

I don't have a strong opinion about this, but I think a wildcard subscription should always stay one (i.e one cannot downgrade it to a regular subscription). If so, and to avoid the problem of an inconsistent state, I think a subscription in "new" wildcard mode should send all its state on stream refresh (alternatively, it should reset its state to the state at the time of creation). 

WDYT? @markdroth?

Reply:
The current spec (without my changes in #16861) says the following:

> For Listener and Cluster resource types, there is also a “wildcard” mode, which is triggered when the initial request on the stream for that resource type contains no resource names. In this case, the server should use site-specific business logic to determine the full set of resources that the client is interested in, typically based on the client’s node identification. Note that once a stream has entered wildcard mode for a given resource type, there is no way to change the stream out of wildcard mode; resource names specified in any subsequent request on the stream will be ignored.

And later:

> Note that for Listener and Cluster resource types where the stream is in “wildcard” mode (see How the client specifies what resources to return for details), the set of resources being subscribed to is determined by the server instead of the client, so there is no mechanism for the client to unsubscribe from resources.

In other words, currently, wildcard mode for a given resource type is determined by the very first request for that resource type on the stream, and the stream stays in that mode forever -- it is not possible to later change a stream from wildcard mode to non-wildcard mode or vice-versa.  If that first request has `resource_names` unset, then the stream is in wildcard mode for that resource type, and the `resource_names` field is ignored in all subsequent requests for that resource type.  However, if the first request has `resource_names` set, then the stream is in non-wildcard mode for that resource type, and if there is later a request for that resource type with `resource_names` unset, that means to unsubscribe from all resources of that type; it does *not* mean to enter wildcard mode.

One way to think about the change we're making here is that we want to allow *exiting* wildcard mode after the initial request, but it will still not be possible to *enter* wildcard mode.  (Entering wildcard mode later would break existing gRPC clients, which will send a request with `resource_names` unset to indicate unsubscribing to all resources of that type.)

Although I think a better way to think about it is what I described in #16861: Instead of wildcard being a mode for the stream as a whole, it's now going to be a special type of subscription (`*`), and the existing syntax of leaving `resource_names` unset will just be a syntactic alias for that special subscription.

Reply:
One note about the wildcard mode:

> One way to think about the change we're making here is that we want to allow _exiting_ wildcard mode after the initial request, but it will still not be possible to _enter_ wildcard mode. (Entering wildcard mode later would break existing gRPC clients, which will send a request with `resource_names` unset to indicate unsubscribing to all resources of that type.)

I actually implemented also reentering into the wildcard subscription in this PR, but the reentering can only be done by sending "*". I imagined the current workflow in general terms (or in terms of SotW) to be like (think of the graphs below as deterministic finite state automata that start at "stream init"):
![wildcard-flow](https://user-images.githubusercontent.com/96081/121226251-a12c2900-c88a-11eb-9c84-17b566b12d8a.png)

And for incremental (delta) wildcard workflow would look like this:
![wildcard-flow-incremental](https://user-images.githubusercontent.com/96081/121226324-b86b1680-c88a-11eb-9cdc-033137458e1e.png)

Reply:
Meh, there are typos like "unsubscribe to" where it should be "unsubscribe from", but hopefully those graphs help.

Reply:
In the SotW diagram, why do you go from wildcard to non-wildcard mode when "empty or non-empty resources list"?  If it's empty, then it seems like you should stay in wildcard mode.

As I said above, I don't think it actually makes sense to model this as a "mode" anymore, since that implies that it's state associated with the stream.  I think the right way to think about this is just that there are a set of subscriptions on the stream, and if the initial request contains an empty set, then that's syntactic sugar for the set containing only `*`.  In other words, the first line of code in the request handler on the server would be something like this for SotW:

```
if (is_first_request && request.resource_names().empty()) {
  request.add_resource_names("*");
}
// ...then do normal processing, handling "*" as a wildcard subscription...
```

Or this for incremental:

```
if (is_first_request && request.resource_names_subscribe().empty()) {
  request.add_resource_names_subscribe("*");
}
// ...then do normal processing, handling "*" as a wildcard subscription...
```

In this view, there is no mode, and therefore no need for a state diagram.

Reply:
> In the SotW diagram, why do you go from wildcard to non-wildcard mode when "empty or non-empty resources list"? If it's empty, then it seems like you should stay in wildcard mode.

Hm, good question. I thought that empty resources list was a special case only for the initial request, not for the subsequent requests. My thinking was that before introducing `*` as wildcard subscription, sending any subsequent requests after the initial empty request was sent made no sense. That's because sending another empty request would change nothing, and sending another request with non-empty resource list would be ignored anyway. So after introducing `*`, I thought that any subsequent requests would need to send `*` to keep the wildcard subscription. An empty request would mean unsubscribing from everything, also from the wildcard subscription. So my question would be: before adding `*`, why would we, after sending an initial empty request, send another empty request? I mean, I'm interested in the answer from rather theoretic point of view, as this PR does not implement handling `*` for SotW anyway.

> 
> As I said above, I don't think it actually makes sense to model this as a "mode" anymore, since that implies that it's state associated with the stream. I think the right way to think about this is just that there are a set of subscriptions on the stream, and if the initial request contains an empty set, then that's syntactic sugar for the set containing only `*`. In other words, the first line of code in the request handler on the server would be something like this for SotW:
> 
> ```
> if (is_first_request && request.resource_names().empty()) {
>   request.add_resource_names("*");
> }
> // ...then do normal processing, handling "*" as a wildcard subscription...
> ```

Right, the thinking of modes seems to be pretty stuck with me. I could say that instead of thinking about wildcard and non-wildcard mode, I could think of them as "has wildcard subscription" and "has no wildcard subscription", respectively. Those seem to be equivalent ways of thinking, but yours makes thing simpler. But looking at your logic, it seems to implement what my graph is actually showing - sending subsequent empty request unsubscribes from everything. Also from the wildcard subscription.

> 
> Or this for incremental:
> 
> ```
> if (is_first_request && request.resource_names_subscribe().empty()) {
>   request.add_resource_names_subscribe("*");
> }
> // ...then do normal processing, handling "*" as a wildcard subscription...
> ```
> 
> In this view, there is no mode, and therefore no need for a state diagram.

Reply:
> My thinking was that before introducing `*` as wildcard subscription, sending any subsequent requests after the initial empty request was sent made no sense. That's because sending another empty request would change nothing, and sending another request with non-empty resource list would be ignored anyway. So after introducing `*`, I thought that any subsequent requests would need to send `*` to keep the wildcard subscription. An empty request would mean unsubscribing from everything, also from the wildcard subscription. So my question would be: before adding `*`, why would we, after sending an initial empty request, send another empty request? I mean, I'm interested in the answer from rather theoretic point of view, as this PR does not implement handling `*` for SotW anyway.

There are always multiple requests on a stream, because even if the client has no intention of changing the set of resources it is subscribed to, it still needs to ACK or NACK each response sent to it by the server.

So, yes, there will be additional requests on the stream with empty resource names, and that should *not* mean to stop handling the wildcard.  So I need to amend the code snippet I gave earlier; instead of `is_first_request`, it should check `!seen_non_empty_resource_list_on_stream`.  In other words, an empty resource list is syntactic shorthand for subscribing to `*` only until the server sees a non-empty resource list from the client; as soon as that happens, then an empty resource list means to unsubscribe from all resources instead of meaning a wildcard.

Reply:
> > My thinking was that before introducing `*` as wildcard subscription, sending any subsequent requests after the initial empty request was sent made no sense. That's because sending another empty request would change nothing, and sending another request with non-empty resource list would be ignored anyway. So after introducing `*`, I thought that any subsequent requests would need to send `*` to keep the wildcard subscription. An empty request would mean unsubscribing from everything, also from the wildcard subscription. So my question would be: before adding `*`, why would we, after sending an initial empty request, send another empty request? I mean, I'm interested in the answer from rather theoretic point of view, as this PR does not implement handling `*` for SotW anyway.
> 
> There are always multiple requests on a stream, because even if the client has no intention of changing the set of resources it is subscribed to, it still needs to ACK or NACK each response sent to it by the server.

Alright, thanks for making this clear to me.

> 
> So, yes, there will be additional requests on the stream with empty resource names, and that should _not_ mean to stop handling the wildcard. So I need to amend the code snippet I gave earlier; instead of `is_first_request`, it should check `!seen_non_empty_resource_list_on_stream`. In other words, an empty resource list is syntactic shorthand for subscribing to `*` only until the server sees a non-empty resource list from the client; as soon as that happens, then an empty resource list means to unsubscribe from all resources instead of meaning a wildcard.

I would say that this is more or less the logic I implemented for delta variant. `is_wildcard && !seen_non_empty_resource_list_on_stream` is `WildcardMode::Implicit`, `is_wildcard && seen_non_empty_resource_list_on_stream` is `WildcardMode::Explicit` and `!is_wildcard` is `WildcardMode::Disabled`. The only difference is that I update this `WildcardMode` on subscription interest update, not when creating/sending the request. Not sure if this difference is a problem, though.

Reply:
The set of subscriptions that a client sends should not be affected by the set of resources that the server responds with.  If a client is subscribed to `*` and the server sends back resources A, B, and C in response, the client is still subscribed to `*`, not to A, B, and C.  So in the first request after a stream restart, the client should still subscribe to `*`, not to A, B, and C.

I am not familiar with Envoy's xDS client code, but given that we're modeling a wildcard as an individual subscription and not as a mode for the stream, I'm not sure having `WildcardMode` as a state actually makes sense.  Instead, you might want to consider just doing something like this when constructing the xDS request:

```
// ...construct request as usual...
if (!request.resource_names().empty()) stream.has_sent_non_empty_resource_list_ = true;
if (stream.has_sent_non_empty_resource_list_ && client_is_subscribed_to_wildcard) {
  request.add_resource_names("*");
}
// ...send request on stream as usual...
```

I'm not sure exactly what state you're talking about tracking here for the resources sent by the server.  Is the concern here that you need to delete any resources not resent by the server after the stream restart?  Note that for the resource types that support wildcard subscriptions, the server is required to send back all resources that the client is subscribed to in every response, so you can presumably just wait for the server's initial response on the new stream and then delete any resource that was not present.  Or are you talking about some different kind of state here?

Reply:
>  If a client is subscribed to * and the server sends back resources A, B, and C in response, the client is still subscribed to *, not to A, B, and C. So in the first request after a stream restart, the client should still subscribe to *, not to A, B, and C.

We are on the same page then.

> I'm not sure exactly what state you're talking about tracking here for the resources sent by the server.

It's resources sent by the server in response to a wildcard subscription (as opposed to resources explicitly asked for by the client). 

> Is the concern here that you need to delete any resources not resent by the server after the stream restart?

Not a concern, but rather an observation that the server may send a different list of resources after a refresh. We could either do what you are suggesting (reconcile client state with update from the server), or clear out client state on refresh.

Reply:
I think that a stream restart should not proactively clear any cached resources on the client.  That would be consistent with the behavior we'd see by having a distinct xDS transport protocol layer, as discussed in #13009.

Reply:
I was looking at a stream refresh as a "soft reset" -- resetting subscription states without connection closing/reopening. I see your point re: cache clearing, this would require the server treating refresh as just another update; specifically, should `*` subscription change, the server would need to send added/removed resources (for delta). I wonder if this should be explicitly mentioned in the docs (if its not already).

This makes envoy implementation simpler though, I think.

Reply:
Yeah, I think of a stream closing as a transient failure, in which case the client should continue using whatever resources it had until it establishes a new stream.  Once it does that, then the server can send any changes that may be needed.

I think the docs do cover this.  In [ACK/NACK and resource type instance version](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-resource-type-instance-version), it says:

> Note that the version for a resource type is not a property of an individual xDS stream but rather a property of the resources themselves. If the stream becomes broken and the client creates a new stream, the client’s initial request on the new stream should indicate the most recent version seen by the client on the previous stream. Servers may decide to optimize by not resending resources that the client had already seen on the previous stream, but only if they know that the client is not subscribing to a new resource that it was not previously subscribed to. For example, it is generally safe for servers to do this optimization for wildcard LDS and CDS requests, and it is safe to do in environments where the clients will always subscribe to exactly the same set of resources.

And the [Incremental xDS](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#incremental-xds) section says:

> On reconnect the Incremental xDS client may tell the server of its known resources to avoid resending them over the network by sending them in initial_resource_versions. Because no state is assumed to be preserved from the previous stream, the reconnecting client must provide the server with all resource names it is interested in.

Reply:
> On reconnect the Incremental xDS client may tell the server of its known resources to avoid resending them over the network by sending them in initial_resource_versions. Because no state is assumed to be preserved from the previous stream, the reconnecting client must provide the server with all resource names it is interested in.

Hmm, I don't think that's works for a wildcard subscription though: the client doesn't send its current state for a `*`. If the assumption on the server-side is that the state hasn't been preserved, then we need to clear the cache on connection refresh.

Reply:
The client *does* send the current state of the individual resources.  In the incremental protocol, there are two different sets of fields in the request:
- `resource_names_subscribe`: Indicates the set of resource locators being subscribed to.  This would include `*`.
- `initial_resource_versions`: Indicates the set of specific resources already seen by the client, along with their versions.  This would include A, B, and C (or whatever resources the server had previously sent before the original stream was closed).

The server will use the subscription to `*` to determine the set of resources that the client should be given.  However, if that expands to A, B, and C, and the `initial_resource_versions` field indicates that the client already has the latest version of those resources, then the server does not need to resend them.

Reply:
@markdroth: you are indeed correct re: resource versions.

Reply:
Personally, I'd very much like an explicit wildcard mode. 

when reconnection comes into play, there is a chance that non-wildcard stream could become wildcard: 
1. client asks for A, B, C, it's marked as non-wildcard stream
2. client removed A, B, C, network error happens 
3. client reconnect, not the server would think it's a wildcard stream and begin to send everything to it.

@markdroth

Reply:
The code takes care of this scenario, what happens here at 3 is "client reconnects, but does not send any initial request, so the server shouldn't be sending anything back".

Reply:
@stevenzzzz As @krnowak mentioned, that case should never happen, because if the client is no longer subscribed to any resources when it reconnects, it will not send any request for that resource type on the reconnected stream.

I don't see any value for an explicit wildcard mode.

Reply:
@krnowak It might be a good idea to change the PR title to reflect what it actually does. :)

Reply:
Right, changed the PR title and the commit message.